torch-native pipeline parallelism for big models (#2345)
huggingface/accelerate: docs/source/usage_guides/distributed_inference.md

# Distributed Inference with 🤗 Accelerate

Distributed inference can fall into three brackets:

> 1. Loading an entire model onto each GPU and sending chunks of a batch through each model at a time

Reviewer: Suggestion: "Loading an entire model onto each GPU and sending chunks of a batch through each GPU's model copy at a time"

I think it's a bit clearer this way.

> 2. Load parts of a model onto each GPU and process a single input at one time

Reviewer: ```suggestion
2. Loading parts of a model onto each GPU and processing a single input at one time
```

Reviewer: For consistency with the first item in the list, let's say "loading" instead of "load".

> 3. Load parts of a model onto each GPU and use what is called scheduled Pipeline Parallelism to combine the two prior techniques. 

Reviewer: ```suggestion
3. Loading parts of a model onto each GPU and using what is called scheduled Pipeline Parallelism to combine the two prior techniques. 
```

> 3. Load parts of a model onto each GPU and used what is called scheduled Pipeline Parallelism to combine the two prior techniques. 

Reviewer 2: used --> use

> We're going to go through the first and the last, showcasing how to do each as they are more realistic scenarios.

Reviewer: ```suggestion
We're going to go through the first and the last bracket, showcasing how to do each as they are more realistic scenarios.
```

> ## Sending chunks of inputs automatically to each loaded model

Reviewer: ```suggestion
## Sending chunks of a batch automatically to each loaded model
```

Reviewer: To be consistent with the first item in the list

> This is the most memory-intensive solution, as it requires each GPU keeps a full copy of the model in memory at a given time. 

Reviewer: ```suggestion
This is the most memory-intensive solution, as it requires each GPU to keep a full copy of the model in memory at a given time. 
```

On the first GPU, the prompts will be `["a dog", "a cat"]`, and on the second GPU it will be `["a chicken", "a chicken"]`.
Make sure to drop the final sample, as it will be a duplicate of the previous one.

> ## A more memory-efficient version (experimental)

Reviewer: This is a vague title. Let's rename to something a bit more clear. 
E.g. "Memory-efficient pipeline parallelism (experimental)"

This next part will discuss using *pipeline parallelism*. This is an **experimental** API utilizing the [PiPPy library by PyTorch](https://github.com/pytorch/PiPPy/) as a native solution. 

> The general idea with pipeline parallism is say you have 4 GPUs, and a model big enough it can be *split* on four GPUs using `device_map="auto"`. What this version will do is you can send in 4 inputs at at time (for example here, any amount works) and each model chunk will work on an input, then recieve the next input after the prior chunk finished it making it *much* more efficient **and faster** than the prior version. Here's a visual taken from the PyTorch repository:

Reviewer: ```suggestion
The general idea with pipeline parallelism is: say you have 4 GPUs and a model big enough it can be *split* on four GPUs using `device_map="auto"`. With this method you can send in 4 inputs at a time (for example here, any amount works) and each model chunk will work on an input, then receive the next input once the prior chunk finished, making it *much* more efficient **and faster** than the method described earlier. Here's a visual taken from the PyTorch repository:
```

> The general idea with pipeline parallism is say you have 4 GPUs, and a model big enough it can be *split* on four GPUs using `device_map="auto"`. What this version will do is you can send in 4 inputs at at time (for example here, any amount works) and each models chunk will work on an ainput, then recieve the next input after the prior chunk finished it making it *much* more efficient **and faster** than the prior version. Here's a visual taken from the PyTorch repository:

Reviewer 2: each models chunk --> each model chunk
an ainput --> an input

![PiPPy example](https://camo.githubusercontent.com/681d7f415d6142face9dd1b837bdb2e340e5e01a58c3a4b119dea6c0d99e2ce0/68747470733a2f2f692e696d6775722e636f6d2f657955633934372e706e67)

> To use this with Accelerate, we have created a [model zoo](https://github.com/muellerzr/pippy-device-map-playground/) showcasing a number of different models and situations to do so. In this tutorial we'll take GPT2 however across two gpus.

Reviewer: ```suggestion
To illustrate how you can use this with Accelerate, we have created a [model zoo example](https://github.com/muellerzr/pippy-device-map-playground/) showcasing a number of different models and situations. In this tutorial, we'll show this method for GPT2 across two GPUs.
```

> Before anything, please make sure you have the latest pippy installed by performing:

Reviewer: ```suggestion
Before you proceed, please make sure you have the latest pippy installed by running the following:
```

```bash
pip install torchpippy
```

> We require at least version 0.2.0, please perform `pip show torchpippy` to check this!

Reviewer: ```suggestion
We require at least version 0.2.0. To confirm that you have the correct version, run `pip show torchpippy`.
```

> First we need to create the model on the CPU:

Reviewer: ```suggestion
Start by creating the model on the CPU:
```

```{python}
from transformers import GPT2ForSequenceClassification, GPT2Config

config = GPT2Config()
model = GPT2ForSequenceClassification(config)
model.eval()
```

> Next we need to create some example inputs to use. These help PiPPy trace the model.

Reviewer: ```suggestion
Next you'll need to create some example inputs to use. These help PiPPy trace the model.
```

<Tip warning={true}>
    However you make this example will determine the relative batch size that will be used/passed

>     through the model at a given time, so make sure to remember them!

Reviewer: I'm not sure I understand what "them" refers to in "make sure to remember them"

</Tip>

```{python}
input = torch.randint(
    low=0,
    high=config.vocab_size,
    size=(2, 1024),  # bs x seq_len
    device="cpu",
    dtype=torch.int64,
    requires_grad=False,
)
```

> Next we need to actually perform the tracing and get the model ready. To do so you simply use the [`inference.prepare_pippy`] function and it will fully wrap the model for pipeline parallism automatically:

Reviewer: ```suggestion
Next we need to actually perform the tracing and get the model ready. To do so, use the [`inference.prepare_pippy`] function and it will fully wrap the model for pipeline parallelism automatically:
```

```{python}
from accelerate.inference import prepare_pippy
example_inputs = {"input_ids": input}
model = prepare_pippy(model, example_args=(input,))
```

> <Tip>

Reviewer: ```suggestion
<Tip>

```

Reviewer: Added a newline, so that the content inside the tip would render properly

    There are a variety of parameters you can pass through to `prepare_pippy`:

>     * `split_points` will let you determine where to split the model at. By default we use wherever `device_map="auto" declares

Reviewer: ```suggestion
    * `split_points` lets you determine where to split the model. By default, we use wherever `device_map="auto" declares
```

Reviewer: Declares what? Maybe give an example?

Author: Done!

>     * `num_chunks` can be used to determine how the batch will be split and sent to the model itself (so `num_chunks=1` with four split points/four GPUs would have a naive MP where a single input gets passed between the four layer split points)

Reviewer: ```suggestion
    * `num_chunks` determines how the batch will be split and sent to the model itself (so `num_chunks=1` with four split points/four GPUs will have a naive MP where a single input gets passed between the four layer split points)
```

</Tip>

> From here all that's left is to actually perform the distributed inference!

Reviewer: ```suggestion
From here, all that's left is to actually perform the distributed inference!
```

> <Tip warning={true}>

Reviewer: ```suggestion
<Tip warning={true}>

```

>     When passing in inputs, while using `kwargs` are supported currently those are even *more* experimental, so it's highly recommended to just simply pass inputs in as a tuple of arguments.

Reviewer: ```suggestion
When passing inputs, we highly recommend to pass them in as a tuple of arguments. Using `kwargs` is supported, however, this approach is experimental.
```

>     When passing in inputs, while using `kwargs` are supported currently those are even *more* experimental, so it's highly recommended to just simply pass inptus in as a tuple of arguments.

Reviewer 2: inptus --> inputs

</Tip>

```{python}
args = some_more_arguments
with torch.no_grad():
    output = model(*args)
```

> Afterwards all the data will be on the last GPU, which you can use the [`PartialState`] to find and extract:

Reviewer: ```suggestion
When finished, all the data will be on the last GPU, which you can use the [`PartialState`] to find and extract:
```
